From c2c1983aa6b4855cf77d540f8b0c83fd38c2251c Mon Sep 17 00:00:00 2001
From: Juan Palacios <jpalaciosdev@gmail.com>
Date: Thu, 28 Sep 2023 17:39:17 +0200
Subject: [PATCH] Add Botan 3 compilation support

Botan 3 takes precedence over Botan 2 when both versions are installed in the
system.

Fixes #373.
---
Patch mostly based on https://gitlab.com/corectrl/corectrl/-/commit/0650b7c670fd692697b0e6d63c1fe5124a3c1fe7.

 CMakeLists.txt            |  8 ++++++--
 cmake/FindBotan.cmake     | 22 +++++++++++++++++-----
 src/helper/CMakeLists.txt |  8 ++++++--
 3 files changed, 29 insertions(+), 9 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 521c804..d156cff 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -21,7 +21,7 @@ if(NOT TARGET uninstall)
     COMMAND ${CMAKE_COMMAND} -P ${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake)
 endif()
 
-set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD 20)
 
 option (WITH_DEBUG_INFO "Add debug information (to Debug builds)" OFF)
 
@@ -38,7 +38,11 @@ endif()
 find_package(Qt5 5.9 COMPONENTS Core DBus Quick Charts Widgets Network Svg REQUIRED)
 find_package(Qt5LinguistTools REQUIRED)
 find_package(QuaZip REQUIRED)
-find_package(Botan REQUIRED)
+
+find_package(Botan QUIET)
+if(NOT Botan_FOUND)
+  message(FATAL_ERROR "Botan library not found")
+endif()
 
 find_package(fmt 5 QUIET)
 if(NOT fmt_FOUND)
diff --git a/cmake/FindBotan.cmake b/cmake/FindBotan.cmake
index 99584b3..914c40a 100644
--- a/cmake/FindBotan.cmake
+++ b/cmake/FindBotan.cmake
@@ -16,14 +16,13 @@
 # This file is in the public domain
 
 find_package(PkgConfig REQUIRED)
-pkg_check_modules(Botan botan-2)
 
-if(NOT Botan_FOUND)
-  find_path(Botan_INCLUDE_DIRS NAMES botan/botan.h
-      PATH_SUFFIXES botan-2
+macro(FIND_BOTAN_IN_FS BOTAN_VER)
+  find_path(Botan_INCLUDE_DIRS NAMES botan/types.h
+      PATH_SUFFIXES ${BOTAN_VER}
       DOC "The botan include directory")
 
-  find_library(Botan_LIBRARIES NAMES botan botan-2
+  find_library(Botan_LIBRARIES NAMES botan ${BOTAN_VER}
       DOC "The botan library")
 
   # Use some standard module to handle the QUIETLY and REQUIRED arguments, and
@@ -36,6 +35,19 @@ if(NOT Botan_FOUND)
     set(Botan_INCLUDE_DIR ${Botan_INCLUDE_DIRS} CACHE INTERNAL "")
     set(Botan_FOUND ${Botan_FOUND} CACHE INTERNAL "")
   endif()
+endmacro()
+
+pkg_check_modules(Botan botan-3)
+if(NOT Botan_FOUND)
+  find_botan_in_fs(botan-3)
+endif()
+
+if(NOT Botan_FOUND)
+  pkg_check_modules(Botan botan-2)
+
+  if(NOT Botan_FOUND)
+    find_botan_in_fs(botan-2)
+  endif()
 endif()
 
 mark_as_advanced(Botan_INCLUDE_DIRS Botan_LIBRARIES)
diff --git a/src/helper/CMakeLists.txt b/src/helper/CMakeLists.txt
index 3fe2ace..bd78a6b 100644
--- a/src/helper/CMakeLists.txt
+++ b/src/helper/CMakeLists.txt
@@ -1,6 +1,10 @@
 find_package(PkgConfig REQUIRED)
-find_package(Qt5 5.9 COMPONENTS Core DBus REQUIRED)
-find_package(Botan REQUIRED)
+find_package(Qt5 5.15 COMPONENTS Core DBus REQUIRED)
+
+find_package(Botan QUIET)
+if(NOT Botan_FOUND)
+  message(FATAL_ERROR "Botan library not found")
+endif()
 
 set(DBUS_DATADIR_PREFIX_DIR ${CMAKE_INSTALL_FULL_DATADIR})
 option(INSTALL_DBUS_FILES_IN_PREFIX "Use installation prefix for D-Bus files" OFF)
-- 
2.43.0

